src/event-stream/tokens-contract-handler.ts

```diff
@@ -0,0 +1,505 @@
+      await this.storeFtMetadata(fungibleTokenMetadata);
+    } catch (error) {
+      logger.error('error handling FT contract', error);
+      throw error;
+    }
+  }
+
+  /**
+   * fetch Non Fungible contract metadata
+   */
+  private async handleNftContract() {
+    try {
+      const contractCallTokenId = await this.makeReadOnlyContractCall('get-last-token-id', []);
+      const tokenId = this.checkAndParseUintCV(contractCallTokenId);
+
+      if (tokenId) {
+        const contractCallUri = await this.makeReadOnlyContractCall('get-token-uri', [tokenId]);
+        const uriCV = this.checkAndParseOptionalString(contractCallUri);
+
+        let metadata: NftTokenMetadata = { name: '', description: '', imageUrl: '' };
+        if (uriCV) {
+          metadata = await this.getMetadataFromUri<NftTokenMetadata>(uriCV.data);
+        }
+
+        const nonFungibleTokenMetadata: DbNonFungibleTokenMetadata = {
+          token_uri: uriCV ? uriCV.data : '',
+          name: metadata.name,
+          description: metadata.description,
+          image_uri: metadata.imageUrl ? this.getImageUrl(metadata.imageUrl) : '',
+          image_canonical_uri: metadata.imageUrl || '',
+          contract_id: `${this.contractAddress}.${this.contractName}`,
+        };
+        await this.storeNftMetadata(nonFungibleTokenMetadata);
+      }
+    } catch (error) {
+      logger.error('error: error handling NFT contract', error);
+      throw error;
+    }
+  }
+
+  /**
+   * Helper method for creating http/s url for supported protocols.
+   * URLs with `http` or `https` protocols are returned as-is.
+   * URLs with `ipfs` or `ipns` protocols are returned with as an `https` url
+   * using a public IPFS gateway.
+   */
+  private getFetchableUrl(uri: string): URL {
+    const parsedUri = new URL(uri);
+    if (parsedUri.protocol === 'http:' || parsedUri.protocol === 'https:') return parsedUri;
+    if (parsedUri.protocol === 'ipfs:')
+      return new URL(`${PUBLIC_IPFS}/${parsedUri.host}${parsedUri.pathname}`);
+
+    if (parsedUri.protocol === 'ipns:')
+      return new URL(`${PUBLIC_IPFS}/${parsedUri.host}${parsedUri.pathname}`);
+
+    throw new Error(`Unsupported uri protocol: ${uri}`);
+  }
+
+  private getImageUrl(uri: string): string {
+    // Support images embedded in a Data URL
+    if (new URL(uri).protocol === 'data:') {
+      // const dataUrl = ParseDataUrl(uri);
+      const dataUrl = parseDataUrl(uri);
+      if (!dataUrl) {
+        throw new Error(`Data URL could not be parsed: ${uri}`);
+      }
+      if (!dataUrl.mediaType?.startsWith('image/')) {
+        throw new Error(`Token image is a Data URL with a non-image media type: ${uri}`);
+      }
+      return uri;
+    }
+    const fetchableUrl = this.getFetchableUrl(uri);
+    return fetchableUrl.toString();
+  }
+
+  async performFetch<Type>(url: string): Promise<Type> {
+    const MAX_PAYLOAD_SIZE = 1_000_000; // 1 megabyte
+    const result = await fetch(url, { size: MAX_PAYLOAD_SIZE });
```

Comment:
@asimm241 I saw this ddos vulnerability (making the API server OOM by fetching gigs of data) and patched it here, but need a unit test, probably needs to be a fetch mock